### ISSUE TITLE
Question: Minimum supported Clojure version for matcher-combinators?

### ISSUE THREAD

**Issue by lread** [nubank/matcher-combinators#193]
Hiya, I continue to really enjoy using matcher-combinators, thanks again for an awesome library!

One thing I'm not entirely confident of is the minimum Clojure version required to run matcher-combinators. 

I see that [the current pom](https://repo.clojars.org/nubank/matcher-combinators/3.7.2/matcher-combinators-3.7.2.pom) specifies Clojure 1.11.1:

```xml
    <dependency>
      <groupId>org.clojure</groupId>
      <artifactId>clojure</artifactId>
      <version>1.11.1</version>
    </dependency>
```
And that strongly implies that Clojure 1.11.1 is the minimum recommended version.

If that's the case that's fine, but maybe we should also state it clearly in the README?
Lemme know, can do a quick PR for that.

Why do I ask, you ask? Well, I'm doing some maintenance on pomegranate, and I thought, [woo matcher-combinators would be nice in these tests](https://github.com/clj-commons/pomegranate/issues/152), but then I remembered, oh yeah, pomegranate wants to support back at least Clojure 1.8. And I want to test pomegranate with each supported Clojure version on CI.

If 1.11.1 is not the actual min supported version for matcher-combinators,
and you want to add CI tests against older versions of Clojure, I am happy to help with that.

**Comment by philomates** [comment]
We have a clojure 1.8 profile in `project.clj` but running `lein with-profile 1.8 test` errors due to `clojure.spec.alpha`, which I guess doesn't work on 1.8(?).
Spec isn't super used in matcher-combinators so maybe there is a way around it. But we do use the namespace keyword alias thing (`::`), which is clojure 1.9, right?

**Comment by lread** [comment]
Thanks for the reply @philomates!

I think the `::` syntax has been there since the start:

```Clojure
❯ clj -Sdeps '{:deps {org.clojure/clojure {:mvn/version "1.0.0"}}}'
Clojure 1.0.0-
user=> ::foo
:user/foo
user=> 
```

What is newer is [namespace map syntax introduced in Clojure 1.9](https://clojure.atlassian.net/browse/CLJ-1910).


```Clojure
❯ clj -Sdeps '{:deps {org.clojure/clojure {:mvn/version "1.9.0"}}}' 
Clojure 1.9.0
user=> #::{:foo 1}
#:user{:foo 1}
user=> 
```

And Clojure spec is 1.9+ too.

Although it would be convenient for my current use case for matcher-combinators to be 1.8+ compatible, I'm not suggesting it should be. That decision would be up to the matcher-combinators maintainers.

For examples, kaocha is Clojure 1.9+ compatible and the Cognitect test runner seems to be Clojure 1.8+ compatible. Both are reasonable choices.

What I think would be helpful though is deciding on a min Clojure version, testing against it, and documenting it.

**Comment by philomates** [comment]
I was curious what it would take to get Clojure 1.8 support and ended up with this draft https://github.com/nubank/matcher-combinators/pull/194

It is a bit more verbose and looses some spec checks (which are pretty minimal / unnecessary anyways).
I am curious if there is much Clojure 1.8 usage anymore? Probably makes sense just to leave the code unchanged and set the supported min version to Clojure 1.9. But I guess this is a decision for the Nubank folks, so I'll probably stop poking at this until they chime in.

**Comment by lread** [comment]
> I am curious if there is much Clojure 1.8 usage anymore? 

Yeah, good question. For those that answered the [2022 Clojure Survey](https://www.surveymonkey.com/results/SM-QRiy0fSu3bmDK_2FSNMplVJw_3D_3D/):
![image](https://user-images.githubusercontent.com/967328/214127356-78fe8c15-f838-4835-a587-bc0d371c1aad.png)

Anecdotally, when I took over maintenance of rewrite-clj, I was testing against 1.9+ only, and I had, without really noticing, used some 1.9 specific fns. 
There was a request from a member of Cider tooling team to go back further.
We negotiated supporting back to 1.8+.

There are some folks who would prefer that any ancient project/lib can be started without mod in a REPL. Personally, since Clojure is very strong at forward compatibility, I am totally ok with having to bump the Clojure dependency to current in such cases. 

I discovered it can be an issue where some folks have very strong opinions.
For contentious issues, I look to the Clojure core team for guidance.
My understanding is that they seem to typically test against 1.8+ and typically wont-fix if a bump would do the trick for folks using earlier versions.
